CMakeLists.txt

```diff
@@ -1,16 +1,58 @@
 cmake_minimum_required (VERSION 3.3)
 project (llama)
 
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-set(CMAKE_CXX_STANDARD 11)
+# llama
+find_package(Boost 1.66.0 REQUIRED)
+add_library(${PROJECT_NAME} INTERFACE)
+target_include_directories(${PROJECT_NAME} INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include> $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+target_compile_features(${PROJECT_NAME} INTERFACE cxx_std_11)
+target_link_libraries(${PROJECT_NAME} INTERFACE Boost::headers)
 
-ADD_SUBDIRECTORY("examples")
+# llama::llama to make subdirectory projects work
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
 
+# llama IDE
+file(GLOB_RECURSE llamaSources "${CMAKE_CURRENT_SOURCE_DIR}/include/**")
+add_custom_target("llamaIde" SOURCES ${llamaSources})
+source_group(TREE "${CMAKE_CURRENT_LIST_DIR}/include/llama" FILES ${llamaSources})
+
+# tests
+find_package(Catch2 2.6.1 CONFIG REQUIRED)
```

Comment:
Tests are cool but should the user really be forced to install them?

Reply:
I just verified that (on my fork, which is more advanced, but the CMakeLists.txt is similar) and the tests are not installed:
```
bgruber@dixxi3:/mnt/c/dev/llama/build_l$ sudo make install -j4
[ 20%] Built target llama-treemaptest
[ 20%] Built target llama-simpletest
[ 20%] Built target llama-uidtest
[ 41%] Built target llama-vectoradd
[ 41%] Built target llama-virtualviewtest
[ 51%] Built target llama-alpaka-nbody
[ 72%] Built target llama-alpaka-vectoradd
[ 72%] Built target llama-alpaka-asynccopy
[100%] Built target tests
Install the project...
-- Install configuration: ""
-- Installing: /usr/local/include/llama
-- Installing: /usr/local/include/llama/allocator
-- Installing: /usr/local/include/llama/allocator/SharedPtr.hpp
-- Installing: /usr/local/include/llama/allocator/Stack.hpp
-- Installing: /usr/local/include/llama/allocator/Vector.hpp
-- Installing: /usr/local/include/llama/Array.hpp
-- Installing: /usr/local/include/llama/CompareUID.hpp
-- Installing: /usr/local/include/llama/DatumArray.hpp
-- Installing: /usr/local/include/llama/DatumCoord.dox
-- Installing: /usr/local/include/llama/DatumCoord.hpp
-- Installing: /usr/local/include/llama/DatumStruct.hpp
-- Installing: /usr/local/include/llama/Factory.hpp
-- Installing: /usr/local/include/llama/ForEach.hpp
-- Installing: /usr/local/include/llama/GetCoordFromUID.hpp
-- Installing: /usr/local/include/llama/GetType.hpp
-- Installing: /usr/local/include/llama/GetUID.hpp
-- Installing: /usr/local/include/llama/IntegerSequence.hpp
-- Installing: /usr/local/include/llama/llama.hpp
-- Installing: /usr/local/include/llama/mapping
-- Installing: /usr/local/include/llama/mapping/AoS.hpp
-- Installing: /usr/local/include/llama/mapping/Interface.hpp
-- Installing: /usr/local/include/llama/mapping/One.hpp
-- Installing: /usr/local/include/llama/mapping/SoA.hpp
-- Installing: /usr/local/include/llama/mapping/tree
-- Installing: /usr/local/include/llama/mapping/tree/functor
-- Installing: /usr/local/include/llama/mapping/tree/functor/Idem.hpp
-- Installing: /usr/local/include/llama/mapping/tree/functor/LeafOnlyRT.hpp
-- Installing: /usr/local/include/llama/mapping/tree/functor/MoveRTDown.hpp
-- Installing: /usr/local/include/llama/mapping/tree/GetBlobByte.hpp
-- Installing: /usr/local/include/llama/mapping/tree/GetBlobSize.hpp
-- Installing: /usr/local/include/llama/mapping/tree/Mapping.hpp
-- Installing: /usr/local/include/llama/mapping/tree/MergeFunctors.hpp
-- Installing: /usr/local/include/llama/mapping/tree/operations
-- Installing: /usr/local/include/llama/mapping/tree/operations/ChangeNodeRuntime.hpp
-- Installing: /usr/local/include/llama/mapping/tree/operations/GetNode.hpp
-- Installing: /usr/local/include/llama/mapping/tree/Operations.hpp
-- Installing: /usr/local/include/llama/mapping/tree/Reduce.hpp
-- Installing: /usr/local/include/llama/mapping/tree/toString.hpp
-- Installing: /usr/local/include/llama/mapping/tree/TreeCoord.hpp
-- Installing: /usr/local/include/llama/mapping/tree/TreeElement.hpp
-- Installing: /usr/local/include/llama/mapping/tree/TreeFromDomains.hpp
-- Installing: /usr/local/include/llama/preprocessor
-- Installing: /usr/local/include/llama/preprocessor/macros.hpp
-- Installing: /usr/local/include/llama/Tuple.dox
-- Installing: /usr/local/include/llama/Tuple.hpp
-- Installing: /usr/local/include/llama/Types.hpp
-- Installing: /usr/local/include/llama/TypeTraits.hpp
-- Installing: /usr/local/include/llama/UserDomain.hpp
-- Installing: /usr/local/include/llama/View.hpp
-- Installing: /usr/local/include/llama/VirtualView.hpp
-- Installing: /usr/local/lib/cmake/llama/llama-config.cmake
-- Installing: /usr/local/lib/cmake/llama/llama-config-version.cmake
-- Installing: /usr/local/bin/llama-simpletest
-- Installing: /usr/local/bin/llama-uidtest
-- Installing: /usr/local/bin/llama-treemaptest
-- Installing: /usr/local/bin/llama-vectoradd
-- Installing: /usr/local/bin/llama-virtualviewtest
-- Installing: /usr/local/bin/llama-alpaka-nbody
-- Installing: /usr/local/bin/llama-alpaka-vectoradd
-- Installing: /usr/local/bin/llama-alpaka-asynccopy
bgruber@dixxi3:/mnt/c/dev/llama/build_l$
```

Reply:
Anything to be done here?

Reply:
Ah, sorry, I was unclear (and I overlooked your comment from 4 days ago...). I meant that the test dependency Catch2 was mandatory for building the examples.

Reply:
Ohh thats right Catch2 should only be required if tests will be build.
@bernhardmgruber Why do you not integrated the catch2 as you did it in [mallocMC#179](https://github.com/ComputationalRadiationPhysics/mallocMC/pull/179)

Reply:
After my work on mallocMC and having alpaka as a subtree there, I dislike the idea of keeping copies of dependencies in the repository. It complicates git handling.
Furthermore, some libraries cannot be used directly when I only have the source code to them. So I need to manually create a target now for them as well. Even worse: support for subtree library and installed library. This complicates the build scripts.
E.g. in mallocMC:
```
set(mallocMC_CATCH2_PROVIDER "intern" CACHE STRING "Select which Catch2 is used")
set_property(CACHE mallocMC_CATCH2_PROVIDER PROPERTY STRINGS "intern;extern")
mark_as_advanced(mallocMC_CATCH2_PROVIDER)
if(${mallocMC_CATCH2_PROVIDER} STREQUAL "intern")
    add_library(Catch2::Catch2 INTERFACE IMPORTED)
    target_include_directories(Catch2::Catch2 INTERFACE ${CMAKE_CURRENT_LIST_DIR}/thirdParty/catch2/include)
else()
    find_package(Catch2 CONFIG REQUIRED)
endif()
```
From the developer perspective, all I want to express is that I have a dependency on this library. And with 'find_package(...)` this is one line, as it should be. And this is independent from how the library is provided.

I get the convenience of having dependencies packaged within a checkout. But that's what package managers are for. So you run your `apt install` or `yum install` or `vcpkg install` and you are done.

So back to the original comment: I only find Catch2 and build tests now if `BUILD_TESTING` is turned on.

Reply:
Yes, integrating a dependency is not a good idea imo.

So I tested the recent version. Disabling building the tests works well. :+1:  However I have one **last** wish: When I try to build without catch2 installed, it just throws an error. Imho it should at least give a hint that I can disable the tests or just disable the tests itself if catch2 is not found or (what I would prefer): Disable the tests per default.

However I am fine with any of these three solutions.

Reply:
I added a better error message with instructions when Catch2 is not found.